test/UnitTestGuide.md

```diff
@@ -0,0 +1,648 @@
+
+    `npm run tests:unit -- --filter Candidate`
+
+
+*   To run a specific unit test:
+
+    `npm run tests:unit -- --filter Candidate::testValidatePSCID`
+
+
+*   To run the unit tests in debug mode:
+
+    `npm run tests:unit --debug OR npm run tests:unit:debug`
+
+
+
+This is an example of what will show up when the unit tests are executed:
+```
+PHPUnit 6.5.14 by Sebastian Bergmann and contributors.
+
+Runtime:       PHP 7.2.21
+Configuration: /app/test/phpunit.xml
+
+....................................I..........................  63 / 285 ( 22%)
+............................................................... 126 / 285 ( 44%)
+......................................SSSS..................... 189 / 285 ( 66%)
+............................................................... 252 / 285 ( 88%)
+.................................                               285 / 285 (100%)
+
+Time: 2.54 seconds, Memory: 14.00MB
+
+```
+
+
+`.` → Successful unit test
+
+`I` → Incomplete test
+
+`S` → Skipped test
+
+`F` → Failure
+
+`E` → Error
+
+If any tests produce a failure or error, a big red error message will appear in the output. The details of all incomplete and skipped tests will be listed below the summary section. See “The Basics” section below for more information on incomplete and skipped tests. 
+
+
+
+##### **How does this command work?**
+
+`npm run tests:unit` is a script specified in the package.json file in your LORIS root directory. It is defined as: 
+
+
+`"tests:unit": "./test/dockerized-unit-tests.sh"`
+
+
+So, the script runs the contents of [./test/dockerized-unit-tests.sh](https://github.com/aces/Loris/blob/major/test/dockerized-unit-tests.sh). If we take a look at this file, it runs the unit tests using docker-compose and vendor/bin/phpunit. It specifies which tests to run with this line: 
+
+
+`--configuration test/phpunit.xml --testsuite LorisUnitTests $*`
+
+
+The list of tests to run is defined in [./test/phpunit.xml](https://github.com/aces/Loris/blob/major/test/phpunit.xml) under the “LorisUnitTests” testsuite section. If you look at this testsuite block, you can see that it refers to every file in the ./test/unittests/ directory!
+
+
+### **Troubleshooting**
+**General Errors:**
+
+If  the _major_ branch has been updated on the Loris repo, and your test-dev environment is now out of sync (branch and/or database) you will see seemingly unrelated errors like: 	
+
+Example A:
+
+
+`Fatal error:  Uncaught LorisNoSuchModuleException`
+
+
+Example B:
+
+`E: Package 'mysql-client' has no installation candidate \
+ERROR: Service 'unit-tests' failed to build: The command '/bin/sh -c apt-get update &&     `	
+
+<span style="text-decoration:underline;">How to fix:</span>
+
+You will need to update your environment to the major branch: 
+
+Rebase your branch and update your database. It is convenient to reload a backup of your database and then run the patches needed to update it to _major_, but only do this if nothing else is working!
```

Comment:
```
... but only do this if nothing else is working!
```
What do you mean by this phrase? only rebase and update the DB or just update the DB?
What should we try before the last resort option?

I would like a little bit more clarity.

Reply:
I meant only use a backup of your database if nothing else is working. I'll make this more clear!